proposals/nnnn-inline-spirv-and-decorated-types.md

```diff
@@ -0,0 +1,130 @@
+
+* Proposal: [NNNN](NNNN-filename.md)
+* Author(s): [Cassandra Beckley](https://github.com/cassiebeckley)
+* Status: **Design In Progress**
+
+## Introduction
+
+In this proposal, we define the `spirv.inline.Type` and `spirv.DecoratedType`
+target extension types for the SPIR-V backend.
+
+## Motivation
+
+We would like to implement `SpirvType` and `SpirvOpaqueType` from [Inline
+SPIR-V (HLSL proposal 0011)](https://github.com/microsoft/hlsl-specs/blob/main/proposals/0011-inline-spirv.md#types)
+in the SPIR-V backend. These allow users to create arbitrary `OpType`
+instructions, and use them like any other HLSL type. Using these types, a
+vendor creating a new extension can expose it to users by creating a header
+file without needing to modify the compiler.
+
+Additionally, we need a way to represent types with SPIR-V decorations in LLVM
+IR.
+
+## Proposed solution
+
+### SpirvType
+
+To represent `vk::SpirvType` and `vk::SpirvOpaqueType` in LLVM IR, we will add
+three new target extension types:
+
+* `spirv.inline.Type`
+* `spirv.inline.IntegralConstant`
+* `spirv.inline.Literal`
+
+`IntegralConstant` and `Literal` are used to encode arguments to `Type`, and
+may not be used outside that context. They are necessary because target
+extension types must have all type arguments precede all integer arguments,
+whereas SPIR-V type instructions may have an arbitrary number of type,
+immediate literal, and constant id operands in any order.
+
+#### `spirv.inline.Type`
+
+```
+target("spirv.inline.Type", operands..., opcode)
+```
+
+`opcode` is an integer literal representing the opcode of the `OpType`
+instruction to be generated. `operands` represents a list of type arguments
+encoding the operands of the `OpType` instruction. Each operand must be one of:
+
+* A type argument, which will be lowered to the id of the lowered SPIR-V type
+* A `spirv.inline.IntegralConstant`, which will be lowered to the id of an
+  `OpConstant` instruction
+* A `spirv.inline.Literal`, which will be lowered to an immediate literal value
+
+#### `spirv.inline.IntegralConstant`
+
+```
+target("spirv.inline.IntegralConstant", integral_type, value)
+```
+
+`integral_type` is the type argument for the `OpConstant` instruction to be
+generated, and `value` is its literal integer value.
+
+#### `spirv.inline.Literal`
+
+```
+target("spirv.inline.Literal", value)
+```
+
+`value` is a `spirv.inline.IntegralConstant` type which represents the literal
+integer value to be generated.
+
+#### Example
+
+Here's an example of using these types to represent an array of images:
+
+```
+%type_2d_image = type target("spirv.Image", float, 1, 2, 0, 0, 1, 0)
+%integral_constant_28 = type target("spirv.inline.IntegralConstant", i32, 28)
+%integral_constant_4 = type target("spirv.inline.IntegralConstant", i32, 4)
+%ArrayTex2D = type target("spirv.inline.Type", %type_2d_image, %integral_constant_4, 28)
+```
+
+### Type decorations
```

Comment:
You should mention that this is related to the decoration attribute in inline spir-v, which can only be applied to types and member variables. Talk about how that attribute will be translated into something of this type. I believe this has alternative solutions that we discussed. The might be worth mentioning in the alternative solutions seciton.

Lastly, you need to consider all versions of the decoration, decorateId and decorateString? Do we want multiple versions? Use the IngralConstant, Literal, and some other String class?

Sorry, I should have mentioned these variants when I asked you to include the decorations in this design.

Reply:
Also I had not noticed `int_spv_assign_decoration` in https://github.com/llvm/llvm-project/blob/main/llvm/docs/SPIRVUsage.rst#target-intrinsics. We should explain why we are not using it, if we decide not to use it. If it does work for us, then that could be a better solution.